python/ray/dag/compiled_dag_node.py

```diff
@@ -1129,6 +1169,22 @@ def _preprocess(self) -> None:
                     # Add all readers to the NCCL actors of P2P.
                     nccl_actors_p2p.add(downstream_actor_handle)
 
+        # handle_unused_attributes:
+        unused_attributes = self.find_unused_input_attributes(
+            output_node, input_attributes
+        )
+
+        if unused_attributes:
+            unused_attributes_str = ", ".join(str(key) for key in unused_attributes)
+            input_attributes_str = ", ".join(str(key) for key in input_attributes)
+            unused_phrase = "is unused" if len(unused_attributes) == 1 else "are unused"
+
+            raise ValueError(
+                f"DAG expects input: {input_attributes_str}, "
+                f"but {unused_attributes_str} {unused_phrase}. "
+                "Ensure all accessed inputs from InputNode are connected to the output."
```

Comment:
Let's rephrase to:
Compiled Graph expects input to be accessed using all of attributes {input_attributes_str}, but {unused_attributes_str} {unused_phrase}. Ensure all input attributes are used and contribute to the computation of the Compiled Graph output.